File: ios/PTimer/ExposureCalculator/FilterSourcePlusControl.swift
// Copyright © 2026 Sangwook Han
// SPDX-License-Identifier: Apache-2.0

import PTimerCore
import PTimerKit
import SwiftUI

/// The Plus wheel (Filter Set contract, FILTER-PLUS): the trailing-edge
/// control of the wheel row while fewer than four actual wheels exist.
/// It is NOT an actual filter wheel — it selects the Filter Source the
/// next wheel is created from:
///
/// - tap adds a Standard 0 or Filter Set Empty wheel for the displayed
///   source, exactly once (FILTER-PLUS-003);
/// - vertical drag or fling browses Standard / Filter Sets in inventory
///   order, reporting the candidate name so the parent can show it in
///   the status row; releasing on a different source adds exactly one
///   wheel from that final source, while a return to the starting
///   source adds nothing (FR-1.13). The camera's remembered source
///   changes only when the add succeeds; a refused add shows its
///   reason in the status row and changes nothing;
/// - a stationary long press opens Filter Set management; movement
///   past the stationary tolerance cancels it, and once the drag
///   threshold is crossed browsing has won for the rest of the touch
///   (FILTER-PLUS-001).
///
/// At rest the compact control shows the candidate source's color as a
/// secondary cue. VoiceOver exposes ONE focusable element: its
/// adjustable value steps a transient displayed candidate through the
/// sources without adding or touching the camera's remembered source,
/// Manage Filter Sets is always a named action, and Add — activation
/// and the named action — exists only while the DISPLAYED source can
/// add right now; otherwise the hint carries the reason and the
/// element stays focusable and adjustable so the user can browse away
/// (FILTER-A11Y-001, FILTER-PLUS-004, ND-A11Y-002).
struct FilterSourcePlusControl: View {
    let sources: [FilterSource]
    let selectedSource: FilterSource
    let sourceName: (FilterSource) -> String
    let sourceColor: (FilterSource) -> FilterSetColor?
    let pickerHeight: CGFloat
    /// False while a wheel is moving, reshaping, or touched; adding
    /// waits for quiet. Browsing stays enabled.
    let isInteractionQuiet: Bool
    /// Localized reason a given source cannot add right now (`nil`
    /// when it can). Asked for the DISPLAYED source, so a transient
    /// assistive candidate shows and speaks its own availability.
    let addUnavailabilityText: (FilterSource) -> String?
    /// Adds one wheel for the given source; the view model refuses and
    /// reports when the source cannot add.
    let onAdd: (FilterSource) -> Void
    let onManage: () -> Void
    /// Candidate source while browsing, `nil` when idle. The parent
    /// renders the expanded, non-blocking label.
    let onBrowsingChanged: (FilterSource?) -> Void

    /// Classifies the current touch (tap / long press / browse); `nil`
    /// between touches.
    @State private var gesture: FilterSourcePlusGestureArbiter?
    @State private var longPressTask: Task<Void, Never>?
    @State private var browsingHaptic = UISelectionFeedbackGenerator()
    /// The source an assistive increment / decrement is browsing —
    /// displayed and announced, never persisted. Cleared when the
    /// camera's settled source changes (a successful add) so the
    /// element follows the remembered source again.
    @State private var assistiveCandidate: FilterSource?

    private var browsingIndex: Int? {
        gesture?.phase == .browsing ? gesture?.candidateIndex : nil
    }

    private var settledIndex: Int {
        sources.firstIndex(of: selectedSource) ?? 0
    }

    private var candidateSource: FilterSource {
        if let browsingIndex, sources.indices.contains(browsingIndex) {
            return sources[browsingIndex]
        }
        return displayedSource
    }

    /// What a tap or an assistive activation adds: the assistive
    /// candidate while one is being browsed, else the settled source.
    private var displayedSource: FilterSource {
        if let assistiveCandidate, sources.contains(assistiveCandidate) {
            return assistiveCandidate
        }
        return selectedSource
    }

    private var tint: Color {
        sourceColor(candidateSource).map(Color.filterSet) ?? Color.secondary
    }

    /// The displayed source's reason, or `nil` when it can add.
    private var displayedUnavailabilityText: String? {
        addUnavailabilityText(displayedSource)
    }

    private var isAddEnabled: Bool {
        isInteractionQuiet && displayedUnavailabilityText == nil
    }

    var body: some View {
        RoundedRectangle(cornerRadius: 10, style: .continuous)
            .strokeBorder(tint.opacity(browsingIndex == nil ? 0.55 : 0.9), lineWidth: 1.5)
            .background(
                RoundedRectangle(cornerRadius: 10, style: .continuous)
                    .fill(tint.opacity(browsingIndex == nil ? 0.12 : 0.22))
            )
            .overlay {
                VStack(spacing: 3) {
                    Image(systemName: "chevron.up")
                        .font(.system(size: 7, weight: .bold))
                        .foregroundStyle(tint.opacity(0.7))
                    Image(systemName: "plus")
                        .font(.system(size: 13, weight: .semibold))
                        .foregroundStyle(isAddEnabled ? tint : tint.opacity(0.35))
                    Image(systemName: "chevron.down")
                        .font(.system(size: 7, weight: .bold))
                        .foregroundStyle(tint.opacity(0.7))
                }
            }
            .frame(width: 20, height: pickerHeight)
            .contentShape(Rectangle().inset(by: -12))
            // One gesture decides tap / long press / browse from the
            // touch's travel and duration, so the three never compete:
            // a press that stays within the stationary tolerance is a
            // tap (add) or, held half a second, a long press (manage);
            // travel to the drag threshold cancels the long press and
            // browses sources until release (FILTER-PLUS-001).
            .gesture(pressGesture)
            .onChange(of: selectedSource) { _, _ in
                // A successful add moved the remembered source; the
                // element follows it again.
                assistiveCandidate = nil
            }
            .accessibilityElement(children: .ignore)
            .accessibilityLabel(Text("Add filter"))
            .accessibilityValue(Text(sourceName(displayedSource)))
            .accessibilityHint(Text(displayedUnavailabilityText ?? FilterWheelPresenter.plusAccessibilityHint(sourceName: sourceName(displayedSource))))
            // Add is assistive-visible only while the displayed source
            // can add (ND-A11Y-002): the button trait, the activation,
            // and the named action all follow `isAddEnabled`. The
            // element itself keeps one identity across that change so
            // VoiceOver focus survives stepping onto an unavailable
            // source; activation is therefore guarded rather than
            // removed.
            .accessibilityAction {
                guard isAddEnabled else { return }
                onAdd(displayedSource)
            }
            // The activation handler implies the button trait; take it
            // back while Add is unavailable so the element is not
            // presented as activatable.
            .accessibilityRemoveTraits(isAddEnabled ? [] : .isButton)
            // Stepping browses a transient displayed candidate
            // (FILTER-PLUS-004): nothing is added and the camera's
            // remembered source does not change until an add succeeds.
            .accessibilityAdjustableAction { direction in
                let current = sources.firstIndex(of: displayedSource) ?? settledIndex
                let next: Int
                switch direction {
                case .increment: next = current + 1
                case .decrement: next = current - 1
                @unknown default: return
                }
                guard sources.indices.contains(next) else { return }
                assistiveCandidate = sources[next]
            }
            .accessibilityActions {
                if isAddEnabled {
                    Button("Add filter") { onAdd(displayedSource) }
                }
                Button("Manage Filter Sets", action: onManage)
            }
    }

    private var pressGesture: some Gesture {
        DragGesture(minimumDistance: 0, coordinateSpace: .local)
            .onChanged { value in
                if gesture == nil {
                    gesture = FilterSourcePlusGestureArbiter(settledIndex: settledIndex, sourceCount: sources.count)
                    browsingHaptic.prepare()
                    longPressTask?.cancel()
                    longPressTask = Task { @MainActor in
                        try? await Task.sleep(nanoseconds: UInt64(FilterSourcePlusGestureArbiter.longPressDuration * 1_000_000_000))
                        guard !Task.isCancelled, gesture?.deadlineElapsed() == true else { return }
                        onManage()
                    }
                }
                let wasBrowsing = gesture?.phase == .browsing
                guard gesture?.moved(translation: value.translation) == true else {
                    if gesture?.phase != .pressing {
                        // Past the stationary tolerance: the long press
                        // can no longer complete.
                        longPressTask?.cancel()
                    }
                    return
                }
                longPressTask?.cancel()
                if wasBrowsing {
                    browsingHaptic.selectionChanged()
                }
                let candidate = gesture?.candidateIndex
                onBrowsingChanged(candidate.flatMap { sources.indices.contains($0) ? sources[$0] : nil })
            }
            .onEnded { _ in
                longPressTask?.cancel()
                longPressTask = nil
                let outcome = gesture?.released() ?? .none
                gesture = nil
                onBrowsingChanged(nil)
                switch outcome {
                case .addBrowsed(let index):
                    // The browse settled on a different source: add
                    // exactly one wheel from it. The view model refuses
                    // and reports when it cannot add.
                    if sources.indices.contains(index) {
                        onAdd(sources[index])
                    }
                case .add:
                    // A tap adds the displayed source; when adding is
                    // unavailable the view model shows the reason.
                    onAdd(displayedSource)
                case .managed, .none:
                    break
                }
            }
    }
}

extension Color {
    /// Maps the platform-neutral Filter Set color token to SwiftUI's
    /// system colors so both platforms read the same token.
    static func filterSet(_ token: FilterSetColor) -> Color {
        switch token {
        case .red: return .red
        case .orange: return .orange
        case .yellow: return .yellow
        case .green: return .green
        case .mint: return .mint
        case .teal: return .teal
        case .cyan: return .cyan
        case .blue: return .blue
        case .indigo: return .indigo
        case .purple: return .purple
        case .pink: return .pink
        case .brown: return .brown
        }
    }
}

extension FilterSetColor {
    /// Localized color name — color is never the only identifying cue
    /// (FILTER-SET-005 / FILTER-A11Y-002).
    var localizedName: String {
        switch self {
        case .red: return String(localized: "Red")
        case .orange: return String(localized: "Orange")
        case .yellow: return String(localized: "Yellow")
        case .green: return String(localized: "Green")
        case .mint: return String(localized: "Mint")
        case .teal: return String(localized: "Teal")
        case .cyan: return String(localized: "Cyan")
        case .blue: return String(localized: "Blue")
        case .indigo: return String(localized: "Indigo")
        case .purple: return String(localized: "Purple")
        case .pink: return String(localized: "Pink")
        case .brown: return String(localized: "Brown")
        }
    }
}

File: android/app/src/main/kotlin/com/sangwook/ptimer/app/ui/shooting/FilterSourcePlusControl.kt
// Copyright © 2026 Sangwook Han
// SPDX-License-Identifier: Apache-2.0

package com.sangwook.ptimer.app.ui.shooting

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.gestures.awaitEachGesture
import androidx.compose.foundation.gestures.awaitFirstDown
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Add
import androidx.compose.material.icons.filled.KeyboardArrowDown
import androidx.compose.material.icons.filled.KeyboardArrowUp
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.rememberUpdatedState
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.hapticfeedback.HapticFeedbackType
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.platform.LocalHapticFeedback
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.semantics.CustomAccessibilityAction
import androidx.compose.ui.semantics.ProgressBarRangeInfo
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.customActions
import androidx.compose.ui.semantics.onClick
import androidx.compose.ui.semantics.progressBarRangeInfo
import androidx.compose.ui.semantics.setProgress
import androidx.compose.ui.semantics.stateDescription
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import com.sangwook.ptimer.R
import com.sangwook.ptimer.app.vm.FilterPlusUiState
import com.sangwook.ptimer.app.vm.FilterSourcePlusGestureArbiter
import com.sangwook.ptimer.app.vm.FilterSourceUiOption
import com.sangwook.ptimer.core.exposure.FilterAddUnavailability
import com.sangwook.ptimer.core.exposure.FilterSource
import com.sangwook.ptimer.ui.theme.filterSetColor
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

/** Slim trailing-edge Plus control width (iOS uses a 26 pt ghost column). */
internal val FilterPlusControlWidth = 26.dp

/**
 * The Plus wheel (FILTER-PLUS): the trailing-edge control of the wheel
 * row while fewer than four actual wheels exist. It is NOT an actual
 * filter wheel — it selects the Filter Source the next wheel is created
 * from:
 *
 * - a tap adds a Standard 0 or Filter Set Empty wheel for the displayed
 *   source, exactly once (FILTER-PLUS-003);
 * - a vertical drag browses Standard / Filter Sets in selection order,
 *   reporting the candidate so the status region can show it; releasing
 *   on a different source adds exactly one wheel from that final source,
 *   while a return to the starting source adds nothing;
 * - a stationary long press opens Filter Set management; movement past
 *   the stationary tolerance cancels it, and once the drag threshold is
 *   crossed browsing has won for the rest of the touch (FILTER-PLUS-001).
 *
 * At rest the compact control shows the candidate source's color as a
 * secondary cue. TalkBack sees ONE focusable element (FILTER-A11Y-001,
 * FILTER-PLUS-004): its adjustable value steps a transient displayed
 * candidate through the sources without adding or touching the camera's
 * remembered source, Manage Filter Sets is always a named action, and
 * Add — activation and the named action — exists only while the
 * DISPLAYED source can add right now; otherwise the state description
 * carries the reason and the element stays adjustable so the user can
 * browse away.
 */
@Composable
internal fun FilterSourcePlusControl(
    plus: FilterPlusUiState,
    height: Dp,
    onAdd: (FilterSource) -> Unit,
    onManage: () -> Unit,
    onBrowsingChanged: (FilterSourceUiOption?) -> Unit,
    onAddUnavailability: (FilterSource) -> FilterAddUnavailability?,
    modifier: Modifier = Modifier,
) {
    val sources = plus.sources
    val settledIndex = plus.selectedIndex.coerceIn(0, (sources.size - 1).coerceAtLeast(0))

    // The source an assistive increment / decrement is browsing —
    // displayed and announced, never persisted. Cleared when the
    // camera's settled source changes (a successful add) so the element
    // follows the remembered source again.
    var assistiveCandidate by remember { mutableStateOf<FilterSourceUiOption?>(null) }
    LaunchedEffect(settledIndex, sources) { assistiveCandidate = null }

    // The source a touch is currently browsing; drives the tint only.
    var browsingIndex by remember { mutableStateOf<Int?>(null) }

    val displayed = assistiveCandidate ?: sources.getOrNull(settledIndex)
    val candidate = browsingIndex?.let { sources.getOrNull(it) } ?: displayed
    val unavailability = if (assistiveCandidate == null) {
        plus.addUnavailability
    } else {
        displayed?.source?.let { onAddUnavailability(it) }
    }
    val isAddEnabled = if (assistiveCandidate == null) plus.canAdd else unavailability == null

    val tint = candidate?.color?.let { filterSetColor(it) } ?: MaterialTheme.colorScheme.onSurfaceVariant
    val isBrowsing = browsingIndex != null

    val addLabel = stringResource(R.string.nd_add_filter)
    val manageLabel = stringResource(R.string.filter_manage_sets)
    val displayedName = displayed?.name?.let { localizedSourceName(it) }.orEmpty()
    val unavailabilityText = unavailability?.let { filterAddUnavailabilityText(it) }

    val haptics = LocalHapticFeedback.current
    val scope = rememberCoroutineScope()
    val currentOnAdd by rememberUpdatedState(onAdd)
    val currentOnManage by rememberUpdatedState(onManage)
    val currentOnBrowsingChanged by rememberUpdatedState(onBrowsingChanged)

    Box(
        modifier = modifier
            .width(FilterPlusControlWidth)
            .height(height)
            .clip(RoundedCornerShape(10.dp))
            .background(tint.copy(alpha = if (isBrowsing) 0.22f else 0.12f))
            .border(1.5.dp, tint.copy(alpha = if (isBrowsing) 0.9f else 0.55f), RoundedCornerShape(10.dp))
            // One gesture decides tap / long press / browse from the
            // touch's travel and duration, so the three never compete.
            .pointerInput(sources, settledIndex) {
                awaitEachGesture {
                    val down = awaitFirstDown(requireUnconsumed = false)
                    val arbiter = FilterSourcePlusGestureArbiter(settledIndex, sources.size)
                    var longPress: Job? = scope.launch {
                        delay(FilterSourcePlusGestureArbiter.LONG_PRESS_MILLIS)
                        if (arbiter.deadlineElapsed()) currentOnManage()
                    }
                    try {
                        while (true) {
                            val event = awaitPointerEvent()
                            val change = event.changes.firstOrNull { it.id == down.id } ?: break
                            if (!change.pressed) {
                                change.consume()
                                break
                            }
                            val dx = (change.position.x - down.position.x).toDp().value
                            val dy = (change.position.y - down.position.y).toDp().value
                            val moved = arbiter.moved(dx, dy)
                            if (arbiter.phase != FilterSourcePlusGestureArbiter.Phase.pressing) {
                                longPress?.cancel()
                                longPress = null
                            }
                            if (moved) {
                                browsingIndex = arbiter.candidateIndex
                                haptics.performHapticFeedback(HapticFeedbackType.TextHandleMove)
                                currentOnBrowsingChanged(arbiter.candidateIndex?.let { sources.getOrNull(it) })
                            }
                            change.consume()
                        }
                    } finally {
                        longPress?.cancel()
                    }
                    val outcome = arbiter.released()
                    browsingIndex = null
                    currentOnBrowsingChanged(null)
                    when (outcome) {
                        // The controller refuses and reports when the
                        // source cannot add, so the view never gates it.
                        is FilterSourcePlusGestureArbiter.ReleaseOutcome.AddBrowsed ->
                            sources.getOrNull(outcome.index)?.let { currentOnAdd(it.source) }

                        FilterSourcePlusGestureArbiter.ReleaseOutcome.Add ->
                            displayed?.let { currentOnAdd(it.source) }

                        FilterSourcePlusGestureArbiter.ReleaseOutcome.Managed,
                        FilterSourcePlusGestureArbiter.ReleaseOutcome.None,
                        -> Unit
                    }
                }
            }
            .clearAndSetSemantics {
                contentDescription = addLabel
                stateDescription = unavailabilityText
                    ?.let { "$displayedName, $it" }
                    ?: displayedName
                // Adjustable: stepping browses a transient displayed
                // candidate; nothing is added and the camera's remembered
                // source does not change until an add succeeds.
                progressBarRangeInfo = ProgressBarRangeInfo(
                    current = sources.indexOf(displayed).coerceAtLeast(0).toFloat(),
                    range = 0f..(sources.size - 1).coerceAtLeast(0).toFloat(),
                    steps = (sources.size - 2).coerceAtLeast(0),
                )
                setProgress { target ->
                    val current = sources.indexOf(displayed).coerceAtLeast(0)
                    val next = when {
                        target > current -> current + 1
                        target < current -> current - 1
                        else -> return@setProgress false
                    }
                    val option = sources.getOrNull(next) ?: return@setProgress false
                    assistiveCandidate = option
                    true
                }
                onClick(addLabel) {
                    val source = displayed?.source
                    if (!isAddEnabled || source == null) {
                        false
                    } else {
                        onAdd(source)
                        true
                    }
                }
                customActions = buildList {
                    if (isAddEnabled && displayed != null) {
                        add(CustomAccessibilityAction(addLabel) { onAdd(displayed.source); true })
                    }
                    add(CustomAccessibilityAction(manageLabel) { onManage(); true })
                }
            },
        contentAlignment = Alignment.Center,
    ) {
        Column(
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.spacedBy(1.dp),
        ) {
            ChevronGlyph(Icons.Filled.KeyboardArrowUp, tint)
            Icon(
                Icons.Filled.Add,
                contentDescription = null,
                tint = if (plus.canAdd) tint else tint.copy(alpha = 0.35f),
                modifier = Modifier.size(14.dp),
            )
            ChevronGlyph(Icons.Filled.KeyboardArrowDown, tint)
        }
    }
}

@Composable
private fun ChevronGlyph(icon: ImageVector, tint: Color) {
    Icon(
        icon,
        contentDescription = null,
        tint = tint.copy(alpha = 0.7f),
        modifier = Modifier.size(10.dp),
    )
}
